packages/worker/src/workflow/threaded-vm.ts

```diff
@@ -0,0 +1,138 @@
+import { Workflow, WorkflowCreator, WorkflowCreateOptions } from './interface';
+import { WorkerThreadInput, WorkerThreadRequest } from './workflow-worker-thread/input';
+import { WorkerThreadOutput, WorkerThreadResponse } from './workflow-worker-thread/output';
+
+interface Completion<T> {
+  resolve(value: T): void;
+  reject(error: any): void;
+}
+
+interface ErrorConstructor {
+  new (message: string): Error;
+}
+
+function errorNameToClass(name: string): ErrorConstructor {
+  switch (name) {
+    case 'IllegalStateError':
+      return IllegalStateError;
+    default:
+      return Error;
+  }
+}
+
+export class WorkerThreadClient {
+  requestIdx = 0n;
+  requestIdToCompletion = new Map<BigInt, Completion<WorkerThreadOutput>>();
+
+  constructor(protected readonly workerThread: Worker) {
+    workerThread.on('message', ({ requestId, result }: WorkerThreadResponse) => {
+      const completion = this.requestIdToCompletion.get(requestId);
+      if (completion === undefined) {
+        // TODO: log
+        return;
+      }
+      if (result.type === 'error') {
+        const ctor = errorNameToClass(result.name);
+        const err = new ctor(result.message);
+        err.stack = result.stack;
+        completion.reject(err);
+        return;
+      }
+
+      completion.resolve(result.output);
+    });
+  }
+
+  async send(input: WorkerThreadInput): Promise<WorkerThreadOutput> {
+    const requestId = this.requestIdx++;
+    const request: WorkerThreadRequest = { requestId, input };
+    this.workerThread.postMessage(request);
+    const promise = new Promise<WorkerThreadOutput>((resolve, reject) => {
+      this.requestIdToCompletion.set(requestId, { resolve, reject });
+    });
+    return promise;
+  }
+
+  async destroy(): Promise<void> {
+    await this.send({ type: 'destroy' });
+    // TODO: figure this out
+    this.workerThread.unref();
+  }
+}
+
+export class ThreadedVMWorkflowCreator implements WorkflowCreator {
+  protected workflowThreadIdx = 0;
+  protected runIdToThreadNum = new Map<string, number>();
+
+  static async create(
+    threadPoolSize: number,
+    code: string,
+    isolateExecutionTimeoutMs: number
+  ): Promise<ThreadedVMWorkflowCreator> {
+    const workerThreadClients = Array(threadPoolSize)
+      .fill(0)
+      .map(() => new WorkerThreadClient(new Worker(require.resolve('./workflow-worker-thread'))));
+    await Promise.all(
+      workerThreadClients.map((client) => client.send({ type: 'init', code, isolateExecutionTimeoutMs }))
+    );
+    return new this(workerThreadClients);
+  }
+
+  constructor(protected readonly workerThreadClients: WorkerThreadClient[]) {}
+
+  async createWorkflow(options: WorkflowCreateOptions): Promise<Workflow> {
+    const workflowThreadIdx = this.workflowThreadIdx;
+    const { runId } = options.info;
+    this.runIdToThreadNum.set(runId, workflowThreadIdx);
+    const workflow = await VMWorkflowThreadProxy.create(this.workerThreadClients[workflowThreadIdx], options);
```

Comment:
I agree that we can probably do better than round robin but it's a simple enough default.
I've added these worker threads because creating and deleting v8 context is an expensive operation which blocks the event loop.
From my experiments with isolated vm, we have a bit more workflow cache capacity when using multiple isolates and each worker thread is run in its own isolate.